### PR TITLE
Concatenate dictionary of objects along axis=1

### LINKED ISSUE #15115: [FEA] Allow `cudf.concat` to process a dictionary of frames / add `keys=`
In pandas, you can pass a dictionary of dataframes to construct a MultiIndex dataframe. This is especially useful if the values contain frames that have the same columns, in which case the keys can help differentiate them.

Right now, the following code:
```python
d = {
    'first': cudf.DataFrame({'A': [1, 2], 'B': [3, 4]}),
    'second': cudf.DataFrame({'A': [5, 6], 'B': [7, 8]}),
}

cudf.concat(d, axis=1)
```
raises the error
```none
TypeError: cannot concatenate object of type <class 'str'>
```
If we try to construct a dataframe from the values of `d`:
```python
cudf.concat(d.values(), axis=1)
```
we get:
```none
NotImplementedError: A Column with duplicate name found: A, cuDF doesn't support having multiple columns with same names yet.
```

If we try to pass keys:
```none
cudf.concat(d.values(), keys=d.keys(), axis=1)
```
we get
```none
TypeError: concat() got an unexpected keyword argument 'keys'
```

Right now, I can do one of two things to get around this issue:
- Mangle the column names of each dataframe beforehand (in a loop) to make them unique and concatenate them. Afterwards, work the columns to make it MultiIndex. 
- Concatenate at the cupy level, create MultiIndex from scratch and create a new dataframe object from these.

In pandas, we could do `pd.concat(d, axis=1)` and call it a day. I wish the same was possible in cudf.

### PR DESCRIPTION
**Pull request by amanlai**
Closes #15115.

Unlike `pandas.concat`, `cudf.concat` doesn't work with a dictionary of objects. The following code raises an error.

```python
d = {
    'first': cudf.DataFrame({'A': [1, 2], 'B': [3, 4]}),
    'second': cudf.DataFrame({'A': [5, 6], 'B': [7, 8]}),
}

cudf.concat(d, axis=1)
```

This commit resolves this issue.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/core/reshape.py [hunk 1]
```diff
@@ -351,35 +384,54 @@ def concat(objs, axis=0, join="outer", ignore_index=False, sort=None):
...
+                        name = (name,)
+                    if empty_inner:
+                        df[(k, *name)] = cudf.core.column.column_empty_like(
+                            col, newsize=0
+                        )
+                    else:
+                        df[(k, *name)] = col
+
+            # MultiIndex construction here
+            result_columns = cudf.MultiIndex.from_tuples(df.columns)
```
Comment (shwina) [inline python/cudf/cudf/core/reshape.py:428]: Accessing `._column_names` is much faster than accessing `.columns`. The latter must return a `pd.Index` object, while the former is a list we already have on hand:

```suggestion
            result_columns = cudf.MultiIndex.from_tuples(df._column_names)
```

File: python/cudf/cudf/core/reshape.py [hunk 2]
```diff
@@ -288,6 +308,19 @@ def concat(objs, axis=0, join="outer", ignore_index=False, sort=None):
                 result = cudf.DataFrame._from_data(
                     data, index=obj.index.copy(deep=True)
                 )
+                if keys is not None:
+                    if isinstance(result, cudf.DataFrame):
+                        k = keys[0]
+                        result.columns = cudf.MultiIndex.from_tuples(
+                            [
+                                (k, *c) if isinstance(c, tuple) else (k, c)
+                                for c in result.columns
```
Comment (shwina) [inline python/cudf/cudf/core/reshape.py:317]:
```suggestion
                                for c in result._column_names
```

File: python/cudf/cudf/core/reshape.py [hunk 3]
```diff
@@ -288,6 +308,19 @@ def concat(objs, axis=0, join="outer", ignore_index=False, sort=None):
...
+                        k = keys[0]
+                        result.columns = cudf.MultiIndex.from_tuples(
+                            [
+                                (k, *c) if isinstance(c, tuple) else (k, c)
+                                for c in result._column_names
+                            ]
+                        )
+
+                    result.columns = cudf.MultiIndex.from_product(
+                        [keys, result.columns]
```
Comment (shwina) [inline python/cudf/cudf/core/reshape.py:322]:
```suggestion
                        [keys, result._column_names]
```

**Comment by shwina** [comment]
Thanks @amanlai - I pushed a couple of quick changes to your branch, which simply make it so that we access the `._column_names` attribute rather than `.columns` when trying to get the column names.

The latter can be surprisingly expensive in cuDF and therefore we forbid its use within the code base.

Fingers crossed that we will now pass all tests 🤞 after which I think this should be good to go!

**Comment by shwina** [comment]
@amanlai - after getting rid of the errors from using `.columns`, it looks like we're down to three unit test failures:

```
FAILED test_concat.py::test_concat_dictionary[d1] - NotImplementedError: Unsupported column type passed to create an Index: <class 'cudf.core.column.lists.ListColumn'>
FAILED test_concat.py::test_concat_dictionary[d3] - AssertionError: DataFrame.columns are different
FAILED test_concat.py::test_concat_dict_incorrect_type - AttributeError: 'Index' object has no attribute 'index'
```

Do you want to debug and fix those up? If you're having trouble, let us know and we're glad to help!

**Comment by shwina** [comment]
Hmm, with the latest changes we're seeing a ton of new failures: 

```
2024-03-03T12:20:33.5760738Z == 575 failed, 100874 passed, 2092 skipped, 867 xfailed in 2584.73s (0:43:04) ==
```

Are you able to run the tests locally to debug? It may be helpful to ensure that you have all tests passing locally, rather than waiting on CI runs to report the results. Typically, I do something like:

```
pytest -n <ncpus> cudf/tests/
```

**Comment by amanlai** [comment]
@shwina I had tested them locally and the test cases were all working fine. However, I don't know why but some files were modified (by my system?) and when I pushed the changes I screwed up and did `git add .` which pushed the unnecessary changes as well. I force pushed the previous version of those files but it may have screwed up even more. I'm just not good at git. I'm sorry that I'm taking too much of your time.

**Comment by shwina** [comment]
No worries at all, and no need to apologize! Can you tell me what the changes we should keep since the commit before last?

**Comment by amanlai** [comment]
@shwina I have no idea why so many tests are failing. I checked on my end and there are no errors. I feel like I have spent enough time on this PR. I want to let somebody else take a crack at it. Sorry that I wasted so much of your time. Cheers.

**Pull request closed without merging**